core/sql/sqlcomp/CmpSeabaseDDLrepos.cpp

```diff
@@ -114,43 +123,72 @@ short CmpSeabaseDDL::createRepos(ExeCliInterface * cliInterface)
 	}
 
       if (endXnIfStartedHere(cliInterface, xnWasStartedHere, cliRC) < 0)
-        return -1;
+        goto label_error;
       
     } // for
   
+  cliRC = cliInterface->restoreCQD("TRAF_MAX_CHARACTER_COL_LENGTH");
   return 0;
+
+  label_error:
+   cliRC = cliInterface->restoreCQD("TRAF_MAX_CHARACTER_COL_LENGTH");
+   return -1;
 }
 
 short CmpSeabaseDDL::dropRepos(ExeCliInterface * cliInterface,
                                NABoolean oldRepos,
-                               NABoolean dropSchema)
+                               NABoolean dropSchema,
+                               NABoolean inRecovery)
 {
   Lng32 cliRC = 0;
-
+  NABoolean xnWasStartedHere = FALSE;
   char queryBuf[1000];
 
   for (Int32 i = 0; i < sizeof(allReposUpgradeInfo)/sizeof(MDUpgradeInfo); i++)
     {
       const MDUpgradeInfo &rti = allReposUpgradeInfo[i];
 
+      // If we are dropping the new repository as part of a recovery action,
+      // and there is no "old" table (because the table didn't change in this
+      // upgrade), then don't drop the new table. (If we did, we would be 
+      // dropping the existing data.)
+      if (!oldRepos && inRecovery && !rti.oldName)
+        continue;
+
       if ((oldRepos && !rti.oldName) || (NOT oldRepos && ! rti.newName))
         continue;
 
       str_sprintf(queryBuf, "drop table %s.\"%s\".%s cascade; ",
                   getSystemCatalog(), SEABASE_REPOS_SCHEMA,
                   (oldRepos ? rti.oldName : rti.newName));
-      
+    
+      if (beginXnIfNotInProgress(cliInterface, xnWasStartedHere))
+        {
+          cliInterface->retrieveSQLDiagnostics(CmpCommon::diags());
+          return -1;
+        }    
+
```

Comment:
What are the plans for transaction management for upgrade?  I recently made a bunch of changed to initialize authorization to allow it to run in a single transaction.  Are we going to run upgrade trafodion in multiple transactions?